### PR TITLE
Improve label parsing

### PR DESCRIPTION
**Pull request by jrajahalme**
Fix parsing of labels like 'reservedz=foo".

Use strings.IndexByte() instead of strings.SplitN() when the delimiter is a single character and N==2. This avoids allocating string slices for the SplitN() return values, which has shown up in pprof output.

These changes make `labels.ParseLabel()` about 4x faster than before.

```release-note
Improve label parsing to be 4x faster.
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5928)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/labels/filter.go [hunk 1]
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
 // parseLabelPrefix returns a LabelPrefix created from the string label parameter.
 func parseLabelPrefix(label string) (*LabelPrefix, error) {
 	labelPrefix := LabelPrefix{}
-	t := strings.SplitN(label, ":", 2)
-	if len(t) > 1 {
-		labelPrefix.Source = t[0]
-		labelPrefix.Prefix = t[1]
+	i := strings.IndexByte(label, ':')
+	if i >= 0 {
```
Comment (aanm) [inline pkg/labels/filter.go:89]: shouldn't this be `i > 0`?

File: pkg/labels/filter.go [hunk 2]
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
...
+		labelPrefix.Source = label[:i]
+		labelPrefix.Prefix = label[i+1:]
```
Comment (aanm) [inline pkg/labels/filter.go:91]: This will panic if I execute `parseLabelPrefix(":")`

File: pkg/labels/labels.go [hunk 3]
```diff
@@ -625,13 +621,15 @@ func ParseLabel(str string) *Label {
...
-	lbl.Key = keySplit[0]
-	if len(keySplit) > 1 {
-		if src == LabelSourceReserved && keySplit[0] == "" {
-			lbl.Key = keySplit[1]
+	i := strings.IndexByte(next, '=')
+	if i < 0 {
+		lbl.Key = next
+	} else {
+		if i == 0 && src == LabelSourceReserved {
+			lbl.Key = next[i+1:]
```
Comment (aanm) [inline pkg/labels/labels.go:629]: This line is not covered by a unit test

File: pkg/labels/filter.go [hunk 4]
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
 // parseLabelPrefix returns a LabelPrefix created from the string label parameter.
 func parseLabelPrefix(label string) (*LabelPrefix, error) {
 	labelPrefix := LabelPrefix{}
-	t := strings.SplitN(label, ":", 2)
-	if len(t) > 1 {
-		labelPrefix.Source = t[0]
-		labelPrefix.Prefix = t[1]
+	i := strings.IndexByte(label, ':')
+	if i >= 0 {
+		labelPrefix.Source = label[:i]
```
Comment (aanm) [inline pkg/labels/filter.go:90]: These 2 lines are not covered by a unit test

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/labels/filter.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/labels/filter.go:89]: `IndexByte()` returns non-negative index when the byte is found. `SplitN()` returns 2 if the delimiter is found. This test was `> 1` before, so the corresponding test is `>= 0`. Are you saying that an empty source like ":prefix" should be parsed as `{Source="", Prefix=":prefix"}` instead of `{Source="", Prefix="prefix"}`? If so this would be a bug fix that I'd separate to a commit of it's own.

File: pkg/labels/filter.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/labels/filter.go:91]: In that case `Source = ":"[:0]`, which is `""`, and `Prefix = ":"[1:]`, which is `""`, no panic. I added a test case for `":"` in to TestParseLabel(), which has the equivalent computation to be absolutely sure.

File: pkg/labels/filter.go (hunk 1 above)
Comment (aanm) [reply pkg/labels/filter.go:89]: > `IndexByte()` returns non-negative index when the byte is found. `SplitN()` returns 2 if the delimiter is found. 

I forgot that.

>This test was `> 1` before, so the corresponding test is `>= 0`. Are you saying that an empty source like ":prefix" should be parsed as `{Source="", Prefix=":prefix"}` instead of `{Source="", Prefix="prefix"}`? 

No

File: pkg/labels/labels.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/labels/labels.go:629]: Added a unit test line that should cover this, but not sure if this makes sense:
```
		{"reserved:=key", NewLabel("key", "", LabelSourceReserved)},
```

Comment (aanm) [reply pkg/labels/labels.go:629]: should be `{"reserved:key", NewLabel("key", "", LabelSourceReserved)},`?

File: pkg/labels/filter.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/labels/filter.go:90]: Changed an existing test to cover these lines.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Added more unit tests and a benchmark. Based on this ParseLabels() is now 4x faster than before.

File: pkg/labels/labels.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/labels/labels.go:629]: Test like that already exists, but it does not cover the lines you commented about. The code branch that was not tested before explicitly tests for an empty key AND "reserved" label source.  Note that this PR does not change the behavior of these two lines, only their implementation.

**Review by ianvernon — APPROVED** [review]

File: pkg/labels/filter_test.go [hunk 5]
```diff
@@ -33,7 +33,7 @@ func (s *LabelsPrefCfgSuite) TestFilterLabels(c *C) {
 	}
 
 	dlpcfg := defaultLabelPrefixCfg()
-	d, err := parseLabelPrefix("!ignor[eE]")
+	d, err := parseLabelPrefix(":!ignor[eE]")
```
Comment (aanm) [inline pkg/labels/filter_test.go:36]: Is this change necessary?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/labels/filter_test.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/labels/filter_test.go:36]: It is, to cover the if branch you said is not covered by the unit tests.

Comment (aanm) [reply pkg/labels/filter_test.go:36]: I thought you were accidentally changing behavior.

**Review by aanm — APPROVED** [review]

File: pkg/labels/filter_test.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/labels/filter_test.go:36]: My bad, I should have called out this change in the commit message.
